pkg/util/watch_filters.go

```diff
@@ -0,0 +1,93 @@
+/*
+Copyright 2024 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"context"
+	"fmt"
+
+	machinev1beta1 "github.com/openshift/api/machine/v1beta1"
+	awscapiv1beta1 "sigs.k8s.io/cluster-api-provider-aws/v2/api/v1beta1"
+	capiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// CAPIMachineToMapiMachine maps a Capi Machine to a MAPI machine in the
+// namespace provided.
+func CAPIMachineToMapiMachine(namespace string) func(context.Context, client.Object) []reconcile.Request {
+	return func(ctx context.Context, obj client.Object) []reconcile.Request {
+		klog.V(4).Info(
+			"reconcile triggered by object",
+			"objectType", fmt.Sprintf("%T", obj),
+			"namespace", obj.GetNamespace(),
+			"name", obj.GetName(),
+		)
+
+		return []reconcile.Request{{
+			NamespacedName: client.ObjectKey{Namespace: namespace, Name: obj.GetName()},
+		}}
+	}
+}
+
+// FilterMAPIMachines filters MAPI Machine requests, ensuring they are in the
+// correct namespace.
+func FilterMAPIMachines(namespace string) predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		machine, ok := obj.(*machinev1beta1.Machine)
+		if !ok {
+			panic(fmt.Sprintf("expected to get an of object of type machinev1beta1.Machine: got type %T", obj))
+		}
+
+		// Check namespace
+		return machine.GetNamespace() == namespace
+
+	})
+}
+
+// FilterCAPIMachines filters CAPI Machine requests, ensuring they are in the
+// correct namespace.
+func FilterCAPIMachines(namespace string) predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		machine, ok := obj.(*capiv1beta1.Machine)
+		if !ok {
+			panic(fmt.Sprintf("expected to get an of object of type capiv1beta1.Machine: got type %T", obj))
+		}
+
+		// Check namespace
+		return machine.GetNamespace() == namespace
+
+	})
+}
+
+// FilterAWSMachines filters AWS Machine requests, ensuring they are in the
+// correct namespace.
+func FilterAWSMachines(namespace string) predicate.Predicate {
```

Comment:
Would it be useful to call this FilterInfraMachines and pass in the type (e.g. awscapiv1beta1.AWSMachine) at Predicate building invocation?

Reply:
Yeah, I wasn't 100% happy with this - and I like the idea! I wonder if the go generics could work here too? e.g similar to the new change for watches in controller runtime...